hole-punching-tests/src/main.rs

```diff
@@ -0,0 +1,391 @@
+        .await
+        .context("Failed to connect to relay")?;
+
+    client_setup(&mut swarm, &mut redis, relay_addr.clone(), mode).await?;
+
+    let mut hole_punched_peer_connection = None;
+
+    loop {
+        match (swarm.next().await.unwrap(), hole_punched_peer_connection) {
+            (
+                SwarmEvent::Behaviour(BehaviourEvent::RelayClient(
+                    relay::client::Event::ReservationReqAccepted { .. },
+                )),
+                _,
+            ) => {
+                log::info!("Relay accepted our reservation request.");
+
+                redis
+                    .push(LISTEN_CLIENT_PEER_ID, swarm.local_peer_id())
+                    .await?;
+            }
+            (
+                SwarmEvent::Behaviour(BehaviourEvent::Dcutr(
+                    dcutr::Event::DirectConnectionUpgradeSucceeded {
+                        remote_peer_id,
+                        connection_id,
+                    },
+                )),
+                _,
+            ) => {
+                log::info!("Successfully hole-punched to {remote_peer_id}");
+
+                // Closing the connection to the relay will implicitly close the relayed connection to the other peer.
+                swarm.close_connection(relay_conn_id);
+                hole_punched_peer_connection = Some(connection_id);
+            }
+            (
+                SwarmEvent::Behaviour(BehaviourEvent::Ping(ping::Event {
+                    connection,
+                    result: Ok(rtt),
+                    ..
+                })),
+                Some(hole_punched_connection),
+            ) if mode == Mode::Dial && connection == hole_punched_connection => {
+                println!("{}", serde_json::to_string(&Report::new(rtt))?);
+
+                return Ok(());
+            }
+            (
+                SwarmEvent::Behaviour(BehaviourEvent::Dcutr(
+                    dcutr::Event::DirectConnectionUpgradeFailed {
+                        remote_peer_id,
+                        error,
+                        ..
+                    },
+                )),
+                _,
+            ) => {
+                log::info!("Failed to hole-punched to {remote_peer_id}");
+                return Err(anyhow::Error::new(error));
+            }
+            (SwarmEvent::OutgoingConnectionError { error, .. }, _) => {
+                anyhow::bail!(error)
+            }
+            (
+                SwarmEvent::ConnectionClosed {
+                    connection_id,
+                    cause: Some(error),
+                    ..
+                },
+                _,
+            ) if connection_id == relay_conn_id => {
+                log::warn!("Connection to relay failed: {error}");
+
+                // TODO: Re-connecting is a bit of a hack, we should figure out why the connection sometimes fails.
+                client_setup(&mut swarm, &mut redis, relay_addr.clone(), mode).await?;
```

Comment:
This is a bit of a hack. For some reason, the underlying connection to the _relay_ can sometimes fail during the reservation. I'll create a follow-up ticket to investigate this later but for now, we can just reconnect here.